returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,205 @@
+            results are passed to the iterator-level function.
+        :param output: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        assert (
+            self.seq_ordering == "default"
+        ), f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}"
+        assert (
+            map_seq is not None or map_seq_stream is not None
+        ), "need to either define map_seq or map_seq_stream functions"
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq or self._identity
+        self._map_seq_stream = map_seq_stream or self._identity
+        self._map_outputs = map_outputs
+
+        self._dataset: Optional[Dataset] = None
+        self._data_keys: Optional[List[str]] = None
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._data_iter_seq_idx = -1
+        self._dim_cache: Dict[str, List[Dim]] = {}
+
+        if _meta_info_cache:
+            self._dim_cache = _meta_info_cache["_dim_cache"]
+            self._estimated_num_seqs = _meta_info_cache["_estimated_num_seqs"]
+            self.labels = _meta_info_cache["labels"]
+            self.num_inputs = _meta_info_cache["num_inputs"]
+            self.num_outputs = _meta_info_cache["num_outputs"]
+
+    def initialize(self):
+        """init"""
+        self._lazy_init_num_outputs()
+        super().initialize()
+
+    @property
+    def _meta_info_cache(self) -> Optional[Dict[str, Any]]:
+        if self.num_outputs is None:
+            return None
+        return {
+            "_dim_cache": self._dim_cache,
+            "_estimated_num_seqs": self._estimated_num_seqs,
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
+
+    def _lazy_init_num_outputs(self):
+        if self.num_outputs is not None:
+            return
+
+        dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self.labels = dataset.labels
+        self._estimated_num_seqs = dataset.estimated_num_seqs
+        if self._map_outputs is not None:
+            tdict = TensorDict(self._map_outputs)
+            # TODO: What if feature_dim == None?
+            self.num_inputs = tdict.data["data"].feature_dim.size
+            self.num_outputs = {
+                k: (t.sparse_dim.size, 1) if t.sparse_dim is not None else (t.feature_dim.size, t.ndim)
+                for k, t in tdict.data.items()
+            }
```

Comment:
@albertz I'm not sure how to do the inference properly here. I think the Tensor class w/ Dim behavior v2 leaves the feature_dim often unspecified, but the code seems to want a value here. It's probably not correct to assume `feature_dim = tensor.dims[-1]`, but I'm not sure what to infer otherwise.

Reply:
> I think the Tensor class w/ Dim behavior v2 leaves the feature_dim often unspecified

No. It's just optional but it often is used.

But anyway, for datasets, basically feature dim always either refers to the sparse_dim when it is sparse, or otherwise to the last dim. I think datasets don't really allow anything else currently. The structure is basically always [T] or [T,D]. I'm not really sure what other cases we handle properly. With `get_data_shape`, `is_data_sparse`, `get_data_dim`, it's slightly more flexible.

`num_inputs`/`num_outputs` don't really allow to describe all different cases properly. But that doesn't matter. They shouldn't be used for any critical code, I think. They are very much deprecated.